internal/controller/pullrequest_controller.go

```diff
@@ -186,160 +157,84 @@ func (r *PullRequestReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	default:
-		return nil, nil
+		return nil, fmt.Errorf("unsupported SCM provider: %s", scmProvider.Name)
 	}
 }
 
-func (r *PullRequestReconciler) handleFinalizer(ctx context.Context, pr *promoterv1alpha1.PullRequest, pullRequestProvider scms.PullRequestProvider) (bool, error) {
-	// name of our custom finalizer
-	finalizerName := "pullrequest.promoter.argoporoj.io/finalizer"
+func (r *PullRequestReconciler) handleFinalizer(ctx context.Context, pr *promoterv1alpha1.PullRequest, provider scms.PullRequestProvider) (bool, error) {
+	finalizer := "pullrequest.promoter.argoporoj.io/finalizer"
 
-	var err error
-	// examine DeletionTimestamp to determine if object is under deletion
 	if pr.DeletionTimestamp.IsZero() {
-		// The object is not being deleted, so if it does not have our finalizer,
-		// then lets add the finalizer and update the object. This is equivalent
-		// to registering our finalizer.
-		if !controllerutil.ContainsFinalizer(pr, finalizerName) {
-			err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
-				// TODO: consider skipping the Get on the initial attempt, the object may already be up to date.
-				err = r.Get(ctx, client.ObjectKeyFromObject(pr), pr)
-				if err != nil {
-					return fmt.Errorf("failed to get PullRequest %q: %w", pr.Name, err)
+		if !controllerutil.ContainsFinalizer(pr, finalizer) {
+			return false, retry.RetryOnConflict(retry.DefaultRetry, func() error { //nolint:wrapcheck
+				if err := r.Get(ctx, client.ObjectKeyFromObject(pr), pr); err != nil {
+					return err //nolint:wrapcheck
 				}
-				updated := controllerutil.AddFinalizer(pr, finalizerName)
-				if !updated {
-					// Nothing to update.
-					return nil
+				if controllerutil.AddFinalizer(pr, finalizer) {
+					return r.Update(ctx, pr)
 				}
-				// No need to wrap the error, we'll wrap it outside the retry.
-				return r.Update(ctx, pr)
+				return nil
 			})
-			if err != nil {
-				return false, fmt.Errorf("failed to add finalizer: %w", err)
-			}
 		}
-	} else {
-		// The object is being deleted
-		if controllerutil.ContainsFinalizer(pr, finalizerName) {
-			// our finalizer is present, so lets handle any external dependency
-			err = r.closePullRequest(ctx, pr, pullRequestProvider)
-			if err != nil {
-				// if fail to delete the external dependency here, return with error
-				// so that it can be retried.
-				return false, fmt.Errorf("failed to close pull request %q: %w", pr.Status.ID, err)
-			}
-
-			// remove our finalizer from the list and update it.
-			controllerutil.RemoveFinalizer(pr, finalizerName)
-			if err = r.Update(ctx, pr); err != nil {
-				return true, fmt.Errorf("failed to remove finalizer: %w", err)
-			}
-			return true, nil
+	} else if controllerutil.ContainsFinalizer(pr, finalizer) {
+		if err := r.closePullRequest(ctx, pr, provider); err != nil {
+			return false, fmt.Errorf("failed to close pull request: %w", err)
+		}
+		controllerutil.RemoveFinalizer(pr, finalizer)
```

Comment:
If the following line (update the CR to remove the finalizer) fails, will we fall into an endless loop of trying to close an already-closed PR? Not sure how `closePullRequest` behaves if the PR is already closed.

Could either ensure `closePullRequest` returns a nil error if the PR is already closed or could nest `r.Update` in a retry loop to mitigate failures.